### ISSUE TITLE
Destroy time provisioners not running

### ISSUE THREAD

**Issue by tholu** [hetznercloud/terraform-provider-hcloud#71]
https://www.terraform.io/docs/provisioners/index.html#destroy-time-provisioners

I'm trying to implement a destroy-time provisioner to let the consul-agent running on my nodes leave gracefully before recreating the resource. However, it seems to be ignored.

Does the hcloud provider not support destroy-time provisioners?

I configured it like this:

```
provisioner "remote-exec" {
    when = "destroy"
    inline = [
      "pkill -INT consul",
      "sleep 10" # Give some time to properly leave the cluster
    ]
  }
```

Am I doing something wrong here?

**Comment by LKaemmerling** [comment]
In the documentation (your link) is a notice about this behavior: 



> Destroy-time provisioners can only run if they remain in the configuration at the time a resource is destroyed. If a resource block with a destroy-time provisioner is removed entirely from the configuration, its provisioner configurations are removed along with it and thus **the destroy provisioner won't run**.


The resource will be directly destroyed, so it doesn't run the destroy provisioner. I think this behavior is intended from terraform.

**Comment by tholu** [comment]
@LKaemmerling This does not make any sense. What you quoted describes a specific scenario, in which the destroy provisioner does not run. In general however, it should run - or this feature would be completely pointless.

> Destroy-time provisioners can only run if they remain in the configuration at the time a resource is destroyed. **If a resource block with a destroy-time provisioner is removed entirely from the configuration**, its provisioner configurations are removed along with it and thus **the destroy provisioner won't run.

See the highlighted part above. I'm not removing a resource from the configuration. The scenario that is described applies, when you shut down a part of your cluster and remove it completely from your Terraform configuration. In this case, the nodes in question are destroyed immediately, because the destroy-time provisioner is gone with the configuration of the resource.

This is illustrated in the workaround, that is described:

> To work around this, a multi-step process can be used to safely remove a resource with a destroy-time provisioner:
> * Update the resource configuration to include count = 0.
> * Apply the configuration to destroy any existing instances of the resource, including running the destroy provisioner.
> * Remove the resource block entirely from configuration, along with its provisioner blocks.
> * Apply again, at which point no further action should be taken since the resources were already destroyed.

To summarize, the destroy-time provisioner needs to be implemented for the hcloud provider (as I think this is the level of implementation in Terraform).

**Comment by LKaemmerling** [comment]
After looking into the other provider modules, i can not find something that differs the process of deletion from ours. 

Can you send me your files? I will try to debug it :)

**Comment by tholu** [comment]
@LKaemmerling: Thanks for your feedback! I‘ll try to reproduce in a minimal sample config, perhaps I‘m doing something wrong right now.

**Comment by tholu** [comment]
I created a minimal example (save e.g. as `example.tf`):

```
# Configure the Hetzner Cloud Provider
provider "hcloud" {
  version = "~> 1.8.0"
  token = "<your token>"
}

# Create the example node
resource "hcloud_server" "node" {
  name = "example-node-1"
  image = "debian-9"
  server_type = "cx11"

  ssh_keys = [
    "<your ssh key>",
  ]

  provisioner "remote-exec" {
    inline = [
      "echo 'Hello World!'"
    ]
  }

    provisioner "remote-exec" {
    when = "destroy"
    inline = [
      "echo 'Bye World!'",
      "sleep 10" # Give it some time
    ]
  }
}

output "cluster_ips" {
  value = [
    "${hcloud_server.node.ipv4_address}",
  ]
}
```

The node creates fine with `terraform init && terraform apply`.

The destroy time provisioners run successfully when I do `terraform destroy`.

However, they do **not** run, when I do `terraform taint hcloud_server.node && terraform apply` - which effectively destroys and recreates the resource.

Maybe I'm misunderstanding destroy time provisioners, but in my view, they should definitely run when tainted resources are destroyed & recreated. Perhaps this is an upstream issue.

**Comment by tholu** [comment]
Found one related upstream issue: https://github.com/hashicorp/terraform/issues/14403 - closing here.
